### DISCUSSION TITLE
NVIDIA Shield - “Close all background apps” action

### DISCUSSION THREAD

**Discussion by michael7oliver** [PRProd/HA-Firemote#159] [Q&A] [answered]
This isn’t necessarily a feature request as it is an information request. 

I’d like to be able to run a script on my NVIDIA Shield Pro which closes all background apps. I’ve noticed that I’m able to Customise the Menu Button to do this under the Shield options by selecting “Close all background apps”. 

My hope was that there was a way to call this action from home assistant in a similar way to how your project allows me to call the “find my remote” action. Is someone able to point me in the right direction on this?

Just as background, my reason for this is that one of the live sports apps I use never resets itself to the Home Screen so I’m forever opening it up to a blank “this program is no longer playing screen” and get back to the Home Screen can be quite cumbersome as it recalls the Home Screen from the last time I used the app… typically with last weeks football games!

**Comment by PRProd** [comment, marked as answer]
Hello @michael7oliver 

I've been digging through some of my old research, trying new things, looking up some things, etc. and I think that the best option for your situation might be to use a script that closes just the offending sports app.

I've tried some commands that will close everything running on my NVIDIA Shield Pro, but they aren't always careful about closing only the things that should be closed.  Sometimes they close things that really should stay open which can lead to undesirable results.

I also tried to work out a solution to simulate a double tap or a long press on the remote button, but due to some[ interesting android certification](https://www.reddit.com/r/ShieldAndroidTV/comments/ih81m1/adb_and_keyevent_for_menu_button/) that needed to be done for these units... long story short, that avenue was a dead end too.

My BEST suggestion right now is this:
 -  On your NVIDIA Shield Pro, open the offending "live sports app" that's been giving you a headache, and leave it open
 - In your Home Assistant UI, click on Developer Tools -> States -> Type something into the filter to help you quickly find your Shield device (optional, but helpful) -> Locate the line labeled "app_id", and copy the contents.  In this example, my Shield device is running the YouTube app, and that app's ID happens to be "com.google.android.youtube.tv"
![image](https://github.com/PRProd/HA-Firemote/assets/25486599/2fd2f2ca-bc5c-4fc4-a408-e67b4e8d94eb)
 - Now, we are going to write and test a command to make your Shield properly quit the open sports app.  
   - Click on Services
   - Select "Android Debug Bridge: ADB Command" from the service dropdown
   - Click on "Choose entity" and select your Shield
   - Type ``adb shell am force-stop com.google.android.youtube.tv`` into the Command field, but use the app id you found earlier instead of com.google.android.youtube.tv
   - click the "CALL SERVICE" button
![image](https://github.com/PRProd/HA-Firemote/assets/25486599/6b701ef2-8d2f-485e-8cbb-406b875d779a)
 - If the sports app that you opened in the first step just stopped running on your Shield, then you win!

With this successful step accomplished, you now know what you need to write a home assistant script, if that's what you are interested in doing.  In my example, I created a script called stop_youtube:
```yaml
stop_youtube:
  alias: Stop YouTube
  sequence:
  - service: androidtv.adb_command
    data:
      command: adb shell am force-stop com.google.android.youtube.tv
    target:
      entity_id: media_player.android_tv_192_168_107_209
  mode: single
  icon: mdi:youtube
```

Is that helpful?

-Doug

**Reply by michael7oliver** [reply]
That’s great, thanks so much! I haven’t created the script yet but will get to the tonight. 

Interestingly, I initially thought it wasn’t working as the app (Kayo) was still showing up in my background apps however when I opened the recent apps page and clicked it opened from scratch… I just ran a few tests and it definitely it restarting the app but for whatever reason it still appears in the opened apps. I thought I’d elaborate on this in case anyone else ever picks up on this thread*.

**Reply by michael7oliver** [reply]
Script working too 👍🏻

**Reply by aherbjornsen** [reply]
I just saw this thread and would like to add that there is functionality in the "Android Debug Bridge" integration for closing apps. Just use the "media_player.select_source" service and add an exclamation mark in front.
From the documentation (https://www.home-assistant.io/integrations/androidtv):
```
stop_netflix:
  sequence:
  - service: media_player.select_source
    target:
      entity_id: media_player.fire_tv_living_room
    data:
      source: "!com.netflix.ninja"
```

**Reply by TrialMaster** [reply]
Hi @PRProd & @aherbjornsen 

I've just found this discussion, and had to thank you both so'ooooo much for solving a real pain with the Shield Pro. I have an Unfolded Circle TV Remote which integrates with Home Assistant so I can run the HA Scripts as part of the remote Activities stop sequence.

This is just the best now, thank you!
